### DISCUSSION TITLE
Is it possible to support tree cursor to `goto_previous_sibling`?

### DISCUSSION THREAD

**Discussion by HerringtonDarkholme** [tree-sitter/tree-sitter#2266] [Q&A] [answered]
Hi tree-sitter community! I'm using tree-sitter to build a [command line search app](https://ast-grep.github.io/). Thanks for maintaining this awesome library!

I'm writing a method `prev_all` to extract all sibling nodes before a node. Currently, my implementation requires repeated calls of `prev_sibling`, [link](https://github.com/ast-grep/ast-grep/blob/main/crates/core/src/node.rs#L356-L374). 

I want to optimize this pattern to make full use of `tree_cursor` but looks like it only supports `goto_next_sibling` at the moment. Is it possible to support traversing siblings backward? 

Thanks

**Comment by ahlinc** [comment]
It's possible but requires a lot of work on the C level + Rust & Wasm bindings + tests.

**Reply by HerringtonDarkholme** [reply]
Thanks! So it means no recent plan for shipping this API right? Using `prev_sibling` as a substitute seems a good choice at the moment.

**Reply by ahlinc** [reply]
If to implement the above then to fully support traversing in a backward direction it makes sense to discover an option to add a `goto_last_child()` method also.

**Reply by ahlinc** [reply]
A PR is welcome, I just can suggest a right direction if you have a time and interest to add it.

At this moment I'm focused on another important things on a free time.
Max also mostly focus on another project and add here a new functionality that need there.

**Reply by HerringtonDarkholme** [reply]
I have a question before trying! 
Will adding several new methods bump the MIN_COMPATIBLE_LANGUAGE_VERSION? Or will this require parser library being rebuilt against the new tree-sitter library?

**Reply by ahlinc** [reply]
> Will adding several new methods bump the MIN_COMPATIBLE_LANGUAGE_VERSION?

No.

> Or will this require parser library being rebuilt against the new tree-sitter library?

No, again.

It would be just a series of additional API methods + additional internal functions. That's what I think based on a quick view to the existing code.

**Reply by HerringtonDarkholme** [reply]
Thanks! Do you have any guide to adding `goto_previous_sibling` and `goto_last_child` methods?

**Reply by daumantas-kavolis-sensmetry** [reply, marked as answer]
https://github.com/daumantas-kavolis-sensmetry/tree-sitter/tree/api-extensions but lacks tests

**Reply by HerringtonDarkholme** [reply]
WOW, that's fast... Thank @daumantas-kavolis-sensmetry! @ahlinc would you consider review this change? It seems to add more than two methods of `prev_siblings` though.

**Reply by daumantas-kavolis-sensmetry** [reply]
I had the changes sitting locally for a while but it may take some time before I can add tests for all the new methods.

**Reply by ahlinc** [reply]
I'm impressed! And definitely want to test the `LookaheadIterator` concept by myself.

**Reply by daumantas-kavolis-sensmetry** [reply]
It would be very useful for language servers and generating error messages, together with bidirectional cursors, and completion suggestions. Lookahead iterator already existed internally but was missing public API though it may need to be renamed to something like `SymbolIterator` as it iterates through symbols at a specified parse state, I just went with the internal name.

**Comment by HerringtonDarkholme** [comment]
@daumantas-kavolis-sensmetry may I ask if you want to submit a pull request for this?

**Reply by daumantas-kavolis-sensmetry** [reply]
Not until I have added tests for the new API methods. C API should be mostly fine from my limited manual testing.

**Reply by daumantas-kavolis-sensmetry** [reply]
@HerringtonDarkholme #2324 Fixed some issues and added tests, should be all good now.

**Reply by HerringtonDarkholme** [reply]
OMG it looks fantastic 😍
